hw-event-proxy/proxy.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2021 The Cloud Native Events Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	hwevent "github.com/redhat-cne/sdk-go/pkg/hwevent"
+
+	"github.com/redhat-cne/sdk-go/pkg/channel"
+	"github.com/redhat-cne/sdk-go/pkg/pubsub"
+	"github.com/redhat-cne/sdk-go/pkg/types"
+
+	"github.com/redhat-cne/hw-event-proxy/hw-event-proxy/pb"
+	"github.com/redhat-cne/hw-event-proxy/hw-event-proxy/restclient"
+	"google.golang.org/grpc"
+
+	v1amqp "github.com/redhat-cne/sdk-go/v1/amqp"
+	v1hwevent "github.com/redhat-cne/sdk-go/v1/hwevent"
+	v1pubsub "github.com/redhat-cne/sdk-go/v1/pubsub"
+	log "github.com/sirupsen/logrus"
+)
+
+// SCConfiguration simple configuration to initialize variables
+type SCConfiguration struct {
+	EventInCh chan *channel.DataChan
+	PubSubAPI *v1pubsub.API
+	BaseURL   *types.URI
+}
+
+var (
+	resourceAddress string = "/hw-event"
+	// used by the webhook handlers
+	scConfig  *SCConfiguration
+	pub       pubsub.PubSub
+	eventType string = "HW_EVENT"
+)
+
+// Start hw event plugin to process events,metrics and status, expects rest api available to create publisher and subscriptions
+func Start(wg *sync.WaitGroup, config *SCConfiguration, fn func(e interface{}) error) error { //nolint:deadcode,unused
+	scConfig = config
+
+	// create publisher
+	var err error
+
+	returnURL := fmt.Sprintf("%s%s", config.BaseURL, "dummy")
+	//	pub, err = scConfig.PubSubAPI.CreatePublisher(v1pubsub.NewPubSub(scConfig.BaseURL, resourceAddress))
+	pub, err = scConfig.PubSubAPI.CreatePublisher(v1pubsub.NewPubSub(types.ParseURI(returnURL), resourceAddress))
+	if err != nil {
+		log.Errorf("failed to create a publisher %v", err)
+		return err
+	}
+	log.Infof("Created publisher %v", pub)
+
+	// once the publisher response is received, create a transport sender object to send events.
+	v1amqp.CreateSender(scConfig.EventInCh, pub.GetResource())
+	log.Infof("Created sender %v", pub.GetResource())
+
+	startWebhook()
+
+	return nil
+}
+
+func startWebhook() {
+	http.HandleFunc("/ack/event", ackEvent)
+	http.HandleFunc("/webhook", publishEvent)
+	go func() {
```

Comment:
need retry  on fail .

Reply:
If http server can not start it's a serious issue. Not sure a retry would help? I only see examples with log.Fatal when this happens.

Reply:
what if it fails due to error?

Reply:
go utilwait.Until(func() {
		err := http.ListenAndServe(bindAddress, mux)
		if err != nil {
			utilruntime.HandleError(fmt.Errorf("starting  server failed: %v", err))
		}
	}, 5*time.Second, utilwait.NeverStop)

Reply:
added retry